### ISSUE TITLE
`@sort` directive semantics

### ISSUE THREAD

**Issue by bojanserafimov** [kensho-technologies/graphql-compiler#532]
This is a proposal for semantics of a `@sort` directive. The directive can be used on any property field. It has a direction and optional rank, defined in the schema like this: 
```
directive @sort(direction: String!, rank: Int) on FIELD
```

The `@sort` with rank `0` is done first, and then the `@sort` with rank `1` is used to resolve ties, and so on. If there's only one `@sort`, rank can be omitted. If there's `N` `@sort` directives, they must take all ranks from `0` to `N - 1`.

The `@sort` directive can be used on fields that are not in the output.

Here's an example usage:
```
{
    Animal {
        name @sort(direction: "ascending", rank: 0)
        out_Animal_ParentOf {
            name @sort(direction: "descending", rank: 1)
            uuid @output(out_name: "parent_uuid")
        }
    }
}
```

The compiler frontend sees whether the provided `@sort` directives completely determine the output ordering, and includes this information in the `output_metadata.`

Emitting should be easy to implement, as MATCH, SQL, and Cypher all suport ordering by multiple parameters. I'm not sure about Gremlin, because I couldn't find anything in its documentation.

**Comment by pmantica1** [comment]
Are you sure we can sort without outputting in SQL? I haven't found any examples online where people do so. 

Also, I prefer:
`directive @sort(ascending: Bool!, rank: Int) on FIELD`

If you are using GraphiQL you get autocomplete for the word ascending and you probably get a visible type error if you mistype "true" for "True", (GraphQL uses lowercase true). So overall with this approach there is less of a chance the user makes a mistake when initially making the query.

**Comment by bojanserafimov** [comment]
@pmantica1 yes, you can sort without outputting https://www.db-fiddle.com/f/2JEznqXyqj6VKSnLh5dmV2/0

**Comment by bojanserafimov** [comment]
Good point with `ascending: Bool!`. I don't really have a preference, but I think your suggestion makes sense.

**Comment by obi1kenobi** [comment]
Do `enum` types work in directives? Sort order feels like an enum more than like a boolean.

**Comment by bojanserafimov** [comment]
> Do `enum` types work in directives? Sort order feels like an enum more than like a boolean.

I'll try it out. I was assuming no, because we don't use enums for filter operators.

**Comment by bojanserafimov** [comment]
Enum values work. I parsed a query using an enum value in a directive and then validated it against a schema.

**Comment by obi1kenobi** [comment]
Off-topic, but should we be thinking about switching to enums for filter operators as well? I'm not sure `=` and ones like it are allowed enum values, but maybe there is some opportunity that isn't immediately obvious...

**Comment by bojanserafimov** [comment]
> Off-topic, but should we be thinking about switching to enums for filter operators as well? I'm not sure `=` and ones like it are allowed enum values, but maybe there is some opportunity that isn't immediately obvious...

`=` is not allowed.

Off-off-topic: If we're changing the `@filter` front-end, might be worth thinking about how to express `OR`.

**Comment by pmantica1** [comment]
> Off-topic, but should we be thinking about switching to enums for filter operators as well? I'm not sure `=` and ones like it are allowed enum values, but maybe there is some opportunity that isn't immediately obvious...

I think it'd be pretty good to have since we get autocomplete on filter directives that way and sometimes people don't know which filter directives are already. Though it'd be a pretty major breaking change.

**Comment by bojanserafimov** [comment]
Let's use enums then:
```
enum Order {
    ASCENDING,
    DESCENDING,
}
```

Does the rest of the design sense? Thumbs up this comment if it looks good, so we can start building it.

**Comment by obi1kenobi** [comment]
Big picture, it looks good. A few minor concerns:
- If we call the enum `Order` then that prevents user-defined types from being called that. Let's maybe use something less likely to appear in someone's database, like `OrderByOrder` or maybe even underscore it `_OrderByOrder`?
- Do you think `ASC/DESC` would be sufficiently explanatory while requiring less typing? Not all queries are written in GraphiQL, and a number of databases use `ASC/DESC` as shorthand.
- Thoughts on calling the directive `order_by` instead of `sort`? I believe this is closer to what most databases call this idea, and I'm not familiar with any databases calling it `sort`.

**Comment by bojanserafimov** [comment]
- On obscure enum name: Good point, we should.
- on `ASC/DESC`: No opinion, sounds good.
- on `order_by`: There's pros and cons. I went for `sort` to match the language in our other directives(`@filter`, `@fold`, etc.). In SQL, the syntax is `ORDER BY name`, which reads naturally. In our case it's `name @order_by(...)`, which doesn't read well. I'm guessing this is the reason we called our directive `@filter` instead of `@where`. I have no strong preference.

**Comment by obi1kenobi** [comment]
If it's all the same to you, I have a mild preference for `order_by`. Imagine outputting a list-typed field and seeing `@sort` on it. I am concerned that people might interpret that as sorting the list before outputting it, instead of ordering by that field.

Speaking of lists, what happens if you try to order by a list-typed field? Is that just disallowed, or is there some plan to handle it?
